### PR TITLE
62 update balances store to support individual tokens

### LINKED ISSUE #62: Update balances store to support individual tokens

### PR DESCRIPTION
**Pull request by filipmanole**
# Description

Fixes #62

Fetching Token Ids using ERC165 to check which interface is supported:
- if ERC721Enumerable -> using balanceOf and tokenOfOwnerByIndex methods
- if ERC721 -> getting token ids based on transfer events from address 0x00 to wallet address
- if ERC1155 -> getting token ids based on transfer events from address 0x00 to wallet address

# Steps to test the change

1.  Make sure to have ERC721Enumerable, ERC721 and ERC1155 NFTs in your wallet
2. Set the collection address in `.env` file. 
3. Verify balance stop is updated with the token ids of the NFTs you own4. 
# Checklist:

- [x] I have read and understood [CONTRIBUTING.MD](https://github.com/airswap/airswap-marketplace/blob/main/CONTRIBUTING.md)
- [x] I have included the ticket number in my PR
- [x] I have performed a self-review of my own code
- [x] I have commented the parts where my code is hard to understand

### CODE REVIEW

File: src/redux/stores/balances/balancesApi.ts [hunk 1]
```diff
@@ -1,7 +1,13 @@
 import BalanceChecker from '@airswap/balances/build/contracts/BalanceChecker.json';
 // eslint-disable-next-line import/extensions
 import balancesDeploys from '@airswap/balances/deploys.js';
+import { ADDRESS_ZERO, tokenKinds } from '@airswap/constants';
 import { SwapERC20, Wrapper } from '@airswap/libraries';
+import { abi as ERC165_ABI } from '@openzeppelin/contracts/build/contracts/ERC165.json';
+import { abi as ERC721_ABI } from '@openzeppelin/contracts/build/contracts/ERC721.json';
+// eslint-disable-next-line camelcase
+import { abi as ERC721Enumerable_ABI } from '@openzeppelin/contracts/build/contracts/ERC721Enumerable.json';
+import { abi as ERC1155_ABI } from '@openzeppelin/contracts/build/contracts/ERC1155.json';
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:10]:
```suggestion
import erc165AbiContract from '@openzeppelin/contracts/build/contracts/ERC165.json';
import erc721AbiContract from '@openzeppelin/contracts/build/contracts/ERC721.json';
import erc721EnumerableContract from '@openzeppelin/contracts/build/contracts/ERC721Enumerable.json';
import erc1155Contract from '@openzeppelin/contracts/build/contracts/ERC1155.json';
```

File: src/redux/stores/balances/balancesApi.ts [hunk 2]
```diff
@@ -25,6 +31,12 @@ interface WalletParams {
   tokenAddresses: string[];
 }
 
+interface fetchTokenIdsParams {
+  provider: ethers.providers.Web3Provider;
+  walletAddress: string;
+  collectionTokenAddress: string;
+}
+
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:39]:
```suggestion
```

File: src/redux/stores/balances/balancesApi.ts [hunk 3]
```diff
@@ -91,3 +103,64 @@ export const getTransactionsLocalStorageKey: (
   walletAddress: string,
   chainId: number
 ) => string = (walletAddress, chainId) => `airswap-marketplace/transactions/${walletAddress}/${chainId}`;
+
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:106]:
```suggestion
interface fetchTokenIdsParams {
  provider: ethers.providers.Web3Provider;
  walletAddress: string;
  collectionToken: TokenInfo;
}

```

File: src/redux/stores/balances/balancesApi.ts [hunk 4]
```diff
@@ -91,3 +103,64 @@ export const getTransactionsLocalStorageKey: (
...
+    walletAddress,
+    collectionTokenAddress,
+  }) => {
+    // 0x780e9d63 is the interface ID for erc721 enumerable
+    const ERC721Enumerable = '0x780e9d63';
+
+    const contract = new ethers.Contract(collectionTokenAddress, ERC165_ABI, provider);
+
+    const isERC721Enumerable = await contract.supportsInterface(ERC721Enumerable);
+    if (isERC721Enumerable) {
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:120]:
```suggestion
    const tokenKind = collectionToken.extensions?.kind as TokenKinds;

    if (tokenKind === TokenKinds.ERC721) {
```

Do the same for other tokens

File: src/redux/stores/balances/balancesApi.ts [hunk 5]
```diff
@@ -91,3 +103,64 @@ export const getTransactionsLocalStorageKey: (
...
+      const collectionContract = new ethers.Contract(collectionTokenAddress, ERC721Enumerable_ABI, provider);
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:121]:
```suggestion
      const collectionContract = new ethers.Contract(collectionToken.address, erc721EnumerableContract.abi, provider);
```

Same for other tokens.

File: src/redux/stores/balances/balancesSubscriber.ts [hunk 6]
```diff
@@ -35,5 +37,15 @@ export const configureBalancesSubscriber = () => {
...
       }));
     }
+
+    if (config.collectionToken && !tokenIdsFetched) {
+      tokenIdsFetched = true;
+
+      store.dispatch(fetchTokenIds({
+        provider: library,
+        walletAddress: web3.account,
+        collectionTokenAddress: config.collectionToken,
```
Comment (piersss) [inline src/redux/stores/balances/balancesSubscriber.ts:47]:
```suggestion
        collectionToken: metadata.tokens[config.collectionToken],
```

File: src/redux/stores/balances/balancesApi.ts (hunk 5 above)
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:121]:
```suggestion
    // 0x780e9d63 is the interface ID for erc721 enumerable
    const ERC721Enumerable = '0x780e9d63';

    const tokenKind = collectionToken.extensions?.kind as TokenKinds;

    // @ts-ignore
    if (tokenKind === ERC721Enumerable) {
      const collectionContract = new ethers.Contract(collectionToken.address, erc721EnumerableContract.abi, provider);
```

Ask don to add erc721 enumerable to TokenKinds

File: src/redux/stores/balances/balancesApi.ts [hunk 7]
```diff
@@ -91,3 +103,64 @@ export const getTransactionsLocalStorageKey: (
...
+      const indexes = Array.from({ length: balance }, (_, i) => i);
+
+      const tokenIdsPromises = indexes.map(async index => (await collectionContract.tokenOfOwnerByIndex(walletAddress, BigNumber.from(index))) as BigNumber);
+      const tokenIds = await Promise.all(tokenIdsPromises);
+
+      return tokenIds.map(t => t.toNumber());
+    }
+
+    const isERC721 = await contract.supportsInterface(tokenKinds.ERC721);
+    if (isERC721) {
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:133]:
```suggestion
    if (tokenKind === TokenKinds.ERC721) {
```

File: src/redux/stores/balances/balancesApi.ts [hunk 8]
```diff
@@ -91,3 +103,64 @@ export const getTransactionsLocalStorageKey: (
...
+      const collectionContract = new ethers.Contract(collectionTokenAddress, ERC721_ABI, provider);
+      const transferFilter = collectionContract.filters.Transfer(ADDRESS_ZERO, walletAddress);
+
+      const events = await collectionContract.queryFilter(transferFilter, 0);
+      /* get token ids from past events */
+      const tokenIds: number[] = events.map(e => e.args?.at(2).toNumber());
+
+      /* get unique values */
+      const uniqueTokenIds = tokenIds.filter((element, index) => tokenIds.indexOf(element) === index);
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:142]:
```suggestion
      const uniqueTokenIds = [...new Set(tokenIds)];
```

File: src/redux/stores/balances/balancesApi.ts [hunk 9]
```diff
@@ -91,3 +103,64 @@ export const getTransactionsLocalStorageKey: (
...
+
+      const tokenIdsPromises = indexes.map(async index => (await collectionContract.tokenOfOwnerByIndex(walletAddress, BigNumber.from(index))) as BigNumber);
+      const tokenIds = await Promise.all(tokenIdsPromises);
+
+      return tokenIds.map(t => t.toNumber());
+    }
+
+    const isERC721 = await contract.supportsInterface(tokenKinds.ERC721);
+    if (isERC721) {
+      const collectionContract = new ethers.Contract(collectionTokenAddress, ERC721_ABI, provider);
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:134]:
```suggestion
      const collectionContract = new ethers.Contract(collectionToken.address, erc721AbiContract.abi, provider);
```

File: src/redux/stores/balances/balancesApi.ts [hunk 10]
```diff
@@ -91,3 +103,64 @@ export const getTransactionsLocalStorageKey: (
...
+      /* get unique values */
+      const uniqueTokenIds = tokenIds.filter((element, index) => tokenIds.indexOf(element) === index);
+
+      /* TODO uniqueTokenIds should not be any[] */
+      return uniqueTokenIds;
+    }
+
+    const isERC1155 = await contract.supportsInterface(tokenKinds.ERC1155);
+    if (isERC1155) {
+      const collectionContract = new ethers.Contract(collectionTokenAddress, ERC1155_ABI, provider);
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:150]:
```suggestion
    if (tokenKind === TokenKinds.ERC1155) {
      const collectionContract = new ethers.Contract(collectionToken.address, erc1155Contract.abi, provider);
```

File: src/redux/stores/balances/balancesApi.ts [hunk 11]
```diff
@@ -91,3 +103,64 @@ export const getTransactionsLocalStorageKey: (
...
+      const transferFilter = collectionContract.filters.TransferSingle(null, ADDRESS_ZERO, walletAddress);
+
+      const events = await collectionContract.queryFilter(transferFilter, 0);
+      /* get token ids from past events */
+      const tokenIds: number[] = events.map(e => e.args?.at(3).toNumber());
+
+      /* get unique values */
+      const uniqueTokenIds = tokenIds.filter((element, index) => tokenIds.indexOf(element) === index);
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:158]:
```suggestion
      const uniqueTokenIds = [...new Set(tokenIds)];
```

File: src/redux/stores/balances/balancesApi.ts [hunk 12]
```diff
@@ -91,3 +103,64 @@ export const getTransactionsLocalStorageKey: (
...
+      const transferFilter = collectionContract.filters.Transfer(ADDRESS_ZERO, walletAddress);
+
+      const events = await collectionContract.queryFilter(transferFilter, 0);
+      /* get token ids from past events */
+      const tokenIds: number[] = events.map(e => e.args?.at(2).toNumber());
+
+      /* get unique values */
+      const uniqueTokenIds = tokenIds.filter((element, index) => tokenIds.indexOf(element) === index);
+
+      /* TODO uniqueTokenIds should not be any[] */
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:144]: They are not any[] for me.

File: src/redux/stores/balances/balancesSubscriber.ts [hunk 13]
```diff
@@ -1,26 +1,28 @@
...
+  let tokenIdsFetched = false;
 
   store.subscribe(() => {
-    const { metadata, web3 } = store.getState();
+    const { metadata, web3, config } = store.getState();
 
     const newTokenKeys = Object.keys(metadata.tokens);
     const newTokenKeysString = newTokenKeys.toString();
 
+    if (!web3.chainId || !web3.account) return;
```
Comment (piersss) [inline src/redux/stores/balances/balancesSubscriber.ts:17]:
```suggestion
    if (!web3.chainId || !web3.account || !config.collectionToken) return;
```

File: src/redux/stores/balances/balancesSubscriber.ts [hunk 14]
```diff
@@ -35,5 +37,15 @@ export const configureBalancesSubscriber = () => {
...
+      tokenIdsFetched = true;
+
+      store.dispatch(fetchTokenIds({
+        provider: library,
+        walletAddress: web3.account,
+        collectionTokenAddress: config.collectionToken,
+      }));
+    }
   });
 };
```
Comment (piersss) [inline src/redux/stores/balances/balancesSubscriber.ts:51]:
```suggestion

     store.dispatch(fetchTokenIds({
        provider: library,
        walletAddress: web3.account,
        collectionToken: metadata.tokens[config.collectionToken],
      }));
    }
  });
};
```

File: src/redux/stores/balances/balancesApi.ts (hunk 12 above)
Comment (filipmanole) [reply src/redux/stores/balances/balancesApi.ts]: it was a bug in vs code;

File: src/redux/stores/balances/balancesApi.ts [hunk 15]
```diff
@@ -91,3 +96,87 @@ export const getTransactionsLocalStorageKey: (
...
+  async ({
+    provider,
+    walletAddress,
+    collectionToken,
+  }) => {
+    // 0x780e9d63 is the interface ID for erc721 enumerable
+    const ERC721Enumerable = '0x780e9d63';
+
+    const tokenKind = collectionToken.extensions?.kind as TokenKinds;
+
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:116]:
```suggestion
    const tokenKind = collectionToken.extensions?.kind as TokenKinds;

    // TODO: ERC721 Enumerable does not work yet. Wait for Airswap libraries to support it.
    const ERC721Enumerable = '0x780e9d63';
```

Comment (piersss) [reply src/redux/stores/balances/balancesApi.ts:116]: Asked @dmosites to add support for this.

File: src/redux/stores/balances/balancesApi.ts [hunk 16]
```diff
@@ -91,3 +96,87 @@ export const getTransactionsLocalStorageKey: (
...
+        const addr = await collectionContract.ownerOf(id);
+
+        if (addr === walletAddress) return true;
+
+        return false;
+      });
+
+      const tokenIds = ownedTokenIds.map(t => t.toNumber());
+
+      return tokenIds;
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:152]:
```suggestion
        return addr === walletAddress;
      });

      return ownedTokenIds.map(t => t.toNumber());
```

File: src/redux/stores/balances/balancesApi.ts [hunk 17]
```diff
@@ -91,3 +96,87 @@ export const getTransactionsLocalStorageKey: (
...
+        const balance = (await collectionContract.balanceOf(walletAddress, id)).toNumber();
+
+        if (balance !== 0) return true;
+
+        return false;
+      });
+
+      const tokenIds = ownedTokenIds.map(t => t.toNumber());
+
+      return tokenIds;
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:177]:
```suggestion
        return balance !== 0;
      });

      return ownedTokenIds.map(t => t.toNumber());
```

File: src/redux/stores/balances/balancesApi.ts [hunk 18]
```diff
@@ -91,3 +96,87 @@ export const getTransactionsLocalStorageKey: (
   walletAddress: string,
   chainId: number
 ) => string = (walletAddress, chainId) => `airswap-marketplace/transactions/${walletAddress}/${chainId}`;
+interface fetchTokenIdsParams {
```
Comment (piersss) [inline src/redux/stores/balances/balancesApi.ts:99]:
```suggestion

interface fetchTokenIdsParams {
```

**Review by piersss — APPROVED** [review]
Added some changes and LGTM!

**Pull request merged**